### PR TITLE
refactor: distinguish between placeholder and compat for vnode count

### PR DESCRIPTION
**Pull request by BugenZhao**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

Distinguish between the placeholder value and backward-compatible value for `vnode_count` field in table catalog.

During the process of creating a streaming job, incomplete catalogs will be temporarily synchronized to the frontend nodes, where `vnode_count` remains placeholders. This PR prevents access to `vnode_count` in this scenario, avoiding hitting the backward-compatibility path that could silently and inadvertently return an incorrect value.

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Documentation

- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

If this PR includes changes that directly affect users or other significant modifications relevant to the community, kindly draft a release note to provide a concise summary of these changes. Please prioritize highlighting the impact these changes will have on users.

</details>

### CODE REVIEW

**Comment by BugenZhao** [comment]
* **#18976** <a href="https://app.graphite.dev/github/pr/risingwavelabs/risingwave/18976?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @BugenZhao and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

**Comment by BugenZhao** [comment]
The CI is expected to fail now because the revealed issue will be resolved in an unmerged PR (#18944).

**Comment by BugenZhao** [comment]
> The CI is expected to fail now because the revealed issue will be resolved in an unmerged PR (#18944).

We have now changed the base to #18944, which should pass the CI.

File: src/frontend/src/catalog/table_catalog.rs [hunk 1]
```diff
@@ -634,7 +632,7 @@ impl From<PbTable> for TableCatalog {
                 .map(TableId::from)
                 .collect_vec(),
             cdc_table_id: tb.cdc_table_id,
-            vnode_count: Some(vnode_count), /* from existing (persisted) tables, vnode_count must be set */
+            vnode_count: VnodeCount::set(vnode_count), /* from existing (persisted) tables, vnode_count must be set */
```
Comment (BugenZhao) [inline src/frontend/src/catalog/table_catalog.rs:635]: We obtained the value from `.value()` then construct a `Set` here, which effectively assert that the `maybe_vnode_count` here is not a placeholder value.

**Review by stdrc — APPROVED** [review]
LGTM

**Review by yezizp2012 — APPROVED** [review]
LGTM

**Pull request merged**